### PR TITLE
kubectl-debug: add more debugging profiles

### PR DESCRIPTION
**Pull request by knight42**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

xref https://github.com/kubernetes/kubectl/issues/1108

#### Special notes for your reviewer:

I plan to implement all the remainging debugging profiles listed in https://github.com/kubernetes/kubectl/issues/1108, but I am afraid that the volume of this PR will be too huge, hence I filed a PR with some basic tests to get early feedback from reviewers. If I was on the right track, I would continue to finish my work.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#debugging-profiles
```

### CODE REVIEW

**Comment by knight42** [comment]
@verb Hi! Would you like to have an early review to see if I was on the right track?

**Comment by verb** [comment]
@knight42 sure thing, this is next on my list.

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go [hunk 1]
```diff
@@ -57,3 +70,162 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
 		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileLegacy, target)
 	}
 }
+
+func (p *generalProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	switch target.(type) {
+	case *corev1.Pod:
```
Comment (verb) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:76]: For the case of pod we'll want to detect whether `pod` and `target` are the same and have different behavior. I don't think we plan on stripping labels for debug by ephemeral container.

**Review by verb — CHANGES_REQUESTED** [review]
I think this approach is fine. Note that I need to rebuild context about the implementation, so please point out when I've gotten a detail wrong or made a bad assumption.

Is there anything I can do to help with implementation?

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go (hunk 1 above)
Comment (knight42) [reply staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:76]: hmm I would like to double check, the proposal https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#profile-general says that:
> Probes and labels are stripped from Pod copies to ensure the copy isn't killed and doesn't receive traffic during debugging.

Do we need to strip labels?

Comment (verb) [reply staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:74]: When debugging a pod, there are two outcomes:

1. Creating a copy of the pod. This should strip labels & probes.
2. Adding an ephemeral container to an existing pod. This should not change labels & probes.

I should have been more clear in the KEP. I meant for the section you pasted to apply to "Pod copies" (i.e. the first case above)

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug_test.go [hunk 2]
```diff
@@ -34,6 +34,14 @@ import (
...
 
+func mustNewProfileApplier(profile string) ProfileApplier {
+	applier, err := NewProfileApplier(profile)
+	if err != nil {
+		panic(fmt.Errorf("failed to create profile applier: %s: %v", profile, err))
+	}
+	return applier
+}
+
 func TestGenerateDebugContainer(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (verb) [reply staging/src/k8s.io/kubectl/pkg/cmd/debug/debug_test.go:45]: No, I meant that we should create a `profiles_test.go` to test the methods in `profiles.go` individually, for example `TestGeneralProfileApply`, `TestBaselineProfileApply`, etc. My thinking is that we'll want multiple test cases per profile, so the table in `debug_test.go` would get really long.

**Comment by verb** [comment]
@knight42 are you still working on this?

**Comment by knight42** [comment]
@verb I am still working on it though I might be slow to respond recently due to the busy work 😢

**Comment by sding3** [comment]
@knight42 - are you still working on this ?

I'm looking to use the `kubectl debug` profile that this PR adds, and I wish to avoid having to maintain a fork. So if you aren't currently working on this, I can pick this up.

**Comment by knight42** [comment]
> @knight42 - are you still working on this ?
> 
> I'm looking to use the `kubectl debug` profile that this PR adds, and I wish to avoid having to maintain a fork. So if you aren't currently working on this, I can pick this up.

Hi, I am unavilable to work on this recently, please feel free to take it.

**Pull request closed without merging**